<a href="https://colab.research.google.com/github/VintaBytes/Ciencia-de-datos/blob/main/Ciencia%20De%20Datos%20Con%20Python/CuadernosColab/CienciaDeDatos_Cap%C3%ADtulo025_Resumir_columnas_num%C3%A9ricas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Abrir en Colab"/></a>

# Capítulo 26 — Contar valores y analizar frecuencias

## De resumir números a contar categorías

En el capítulo anterior trabajamos con columnas numéricas.

Calculamos promedios, medianas, valores mínimos, valores máximos y cuartiles. Esos indicadores nos ayudaron a resumir variables como el total de la cuenta, la propina y el tamaño de la mesa.

Pero no todas las columnas se analizan de la misma manera.

En muchas situaciones queremos responder preguntas como:

- ¿qué día aparece más veces en el dataset?
- ¿cuántos registros corresponden al almuerzo y cuántos a la cena?
- ¿qué proporción de las cuentas corresponde a cada categoría?
- ¿hay categorías muy frecuentes o muy poco frecuentes?

Para responder este tipo de preguntas necesitamos contar valores y analizar frecuencias.

## Cargar el dataset de trabajo

Vamos a seguir usando el dataset `tips` de Seaborn.

Mantener el mismo dataset nos permite concentrarnos en una idea nueva: analizar columnas categóricas mediante conteos, proporciones y porcentajes.

In [10]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("tips")

df.head()

,total_bill,tip,sex,smoker,day,time,size
0,16.99,1.01,Female,No,Sun,Dinner,2
1,10.34,1.66,Male,No,Sun,Dinner,3
2,21.01,3.50,Male,No,Sun,Dinner,3
3,23.68,3.31,Male,No,Sun,Dinner,2
4,24.59,3.61,Female,No,Sun,Dinner,4


La tabla muestra las primeras filas del dataset.

Seguimos trabajando con las mismas variables de los capítulos anteriores, pero ahora el foco estará en las columnas categóricas.

Columnas como `sex`, `smoker`, `day` y `time` no se resumen calculando promedios o medianas. En estos casos suele ser más útil contar cuántas veces aparece cada categoría y qué proporción representa dentro del total.

## Identificar columnas categóricas

Para organizar el análisis, vamos a separar las columnas categóricas que nos interesa revisar.

En este dataset vamos a trabajar principalmente con:

- `sex`
- `smoker`
- `day`
- `time`

Cada una de estas columnas permite responder preguntas sobre la composición del dataset.

In [11]:
columnas_categoricas = ["sex", "smoker", "day", "time"]

df[columnas_categoricas].head()

,sex,smoker,day,time
0,Female,No,Sun,Dinner
1,Male,No,Sun,Dinner
2,Male,No,Sun,Dinner
3,Male,No,Sun,Dinner
4,Female,No,Sun,Dinner


La vista reducida muestra solo las columnas categóricas que vamos a analizar.

Cada una contiene valores que representan grupos o categorías. Por ejemplo, `day` indica el día registrado y `time` distingue entre almuerzo y cena.

En este capítulo vamos a empezar contando cuántas veces aparece cada valor en una columna. Para eso usaremos `value_counts()`, una herramienta que ya vimos en capítulos anteriores, pero que ahora vamos a estudiar con más detalle.

## Contar valores con `value_counts()`

Cuando analizamos columnas categóricas, muchas veces queremos saber cuántas veces aparece cada categoría.

A esa cantidad la llamamos **frecuencia**.

Por ejemplo, si el día `Sat` aparece 87 veces, decimos que su frecuencia es 87. Más adelante también vamos a calcular qué porcentaje del total representa esa cantidad.

Una primera pregunta categórica puede ser:

**¿Cuántas cuentas hay registradas para cada día?**

Para responderla vamos a contar cuántas veces aparece cada valor en la columna `day`.

In [12]:
df["day"].value_counts()

,count
day,
Sat,87
Sun,76
Thur,62
Fri,19


La salida muestra la cantidad de cuentas registradas para cada día.

El día con más registros es **sábado (`Sat`)**, con **87 cuentas**. Luego aparecen **domingo (`Sun`)**, con **76**, y **jueves (`Thur`)**, con **62**. El día con menos registros es **viernes (`Fri`)**, con **19 cuentas**.

A estos valores los llamamos **frecuencias absolutas**, porque indican cuántas veces aparece cada categoría.

Por ahora solo estamos contando registros. Esto nos permite describir la composición del dataset, pero no alcanza para afirmar qué día tuvo mayor facturación o mejores propinas.

## Calcular frecuencias relativas

Las frecuencias absolutas indican cuántas veces aparece cada categoría.

Pero a veces no alcanza con saber la cantidad. También puede ser útil saber qué parte del total representa esa cantidad.

A esa relación la llamamos **frecuencia relativa**.

Por ejemplo, si una categoría aparece 87 veces en un dataset de 244 registros, no solo podemos decir que aparece 87 veces: también podemos calcular qué proporción del total representa.

En Pandas, `value_counts()` permite obtener frecuencias relativas usando `normalize=True`.

In [13]:
df["day"].value_counts(normalize=True)

,proportion
day,
Sat,0.356557
Sun,0.311475
Thur,0.254098
Fri,0.077869


La salida muestra las **frecuencias relativas** de la columna `day`.

En lugar de mostrar cantidades de registros, ahora cada valor indica qué parte del total representa cada día.

Por ejemplo:

- `Sat` representa aproximadamente **0.36** del total.
- `Sun` representa aproximadamente **0.31**.
- `Thur` representa aproximadamente **0.25**.
- `Fri` representa aproximadamente **0.08**.

Las frecuencias relativas son útiles porque permiten comparar categorías considerando el total de registros. Sin embargo, expresadas como decimales pueden ser menos intuitivas para leer. Por eso, muchas veces conviene convertirlas en porcentajes.

## Convertir proporciones en porcentajes

Las frecuencias relativas pueden interpretarse como proporciones.

Para hacerlas más fáciles de leer, podemos multiplicarlas por 100 y redondearlas. Así obtenemos el porcentaje de registros que corresponde a cada categoría.

In [14]:
df["day"].value_counts(normalize=True) * 100

,proportion
day,
Sat,35.655738
Sun,31.147541
Thur,25.409836
Fri,7.786885


Ahora las frecuencias relativas aparecen expresadas como porcentajes.

Podemos leer los resultados así:

- `Sat` representa aproximadamente el **35.66%** de los registros.
- `Sun` representa aproximadamente el **31.15%**.
- `Thur` representa aproximadamente el **25.41%**.
- `Fri` representa aproximadamente el **7.79%**.

La información es la misma que en la celda anterior, pero expresada en una escala más familiar. Aun así, la salida todavía puede mejorarse si redondeamos los valores y los organizamos en una tabla junto con las frecuencias absolutas.

## Construir una tabla de frecuencias

Para comunicar mejor el resultado, podemos combinar en una misma tabla:

- la frecuencia absoluta, es decir, la cantidad de registros;
- el porcentaje, es decir, qué parte del total representa cada categoría.

Vamos a construir esa tabla para la columna `day`.

In [15]:
frecuencia_dias = pd.DataFrame({
    "frecuencia": df["day"].value_counts(),
    "porcentaje": df["day"].value_counts(normalize=True) * 100
})

frecuencia_dias["porcentaje"] = frecuencia_dias["porcentaje"].round(2)

frecuencia_dias

,frecuencia,porcentaje
day,,
Sat,87,35.66
Sun,76,31.15
Thur,62,25.41
Fri,19,7.79


La tabla combina dos formas de leer la misma distribución.

La columna `frecuencia` indica cuántas cuentas hay registradas para cada día. La columna `porcentaje` indica qué parte del total representa cada uno.

Podemos observar que:

- `Sat` concentra **87 registros**, equivalentes al **35.66%** del dataset.
- `Sun` concentra **76 registros**, equivalentes al **31.15%**.
- `Thur` concentra **62 registros**, equivalentes al **25.41%**.
- `Fri` concentra **19 registros**, equivalentes al **7.79%**.

Esta tabla es más informativa que el conteo aislado, porque permite ver tanto la cantidad de casos como el peso relativo de cada categoría.

## Aplicar el mismo análisis a otra columna

El mismo procedimiento puede aplicarse a otras columnas categóricas.

Ahora vamos a analizar la columna `time`, que distingue entre almuerzo (`Lunch`) y cena (`Dinner`).

In [16]:
frecuencia_horario = pd.DataFrame({
    "frecuencia": df["time"].value_counts(),
    "porcentaje": df["time"].value_counts(normalize=True) * 100
})

frecuencia_horario["porcentaje"] = frecuencia_horario["porcentaje"].round(2)

frecuencia_horario

,frecuencia,porcentaje
time,,
Dinner,176,72.13
Lunch,68,27.87


La tabla muestra que la mayoría de los registros corresponde a la cena.

En el dataset, `Dinner` aparece con **176 registros**, equivalentes al **72.13%** del total. En cambio, `Lunch` aparece con **68 registros**, equivalentes al **27.87%**.

Esta diferencia es importante para interpretar análisis posteriores. Si más adelante comparamos consumos o propinas por horario, debemos recordar que el dataset no tiene la misma cantidad de registros para almuerzo y cena.

## Analizar otra variable categórica

También podemos revisar la columna `smoker`, que indica si la cuenta corresponde a una mesa registrada como fumadora o no fumadora.

La pregunta ahora es:

**¿Qué proporción de los registros corresponde a cada categoría?**

In [17]:
frecuencia_fumadores = pd.DataFrame({
    "frecuencia": df["smoker"].value_counts(),
    "porcentaje": df["smoker"].value_counts(normalize=True) * 100
})

frecuencia_fumadores["porcentaje"] = frecuencia_fumadores["porcentaje"].round(2)

frecuencia_fumadores

,frecuencia,porcentaje
smoker,,
No,151,61.89
Yes,93,38.11


La tabla muestra la distribución de la columna `smoker`.

En el dataset, la categoría `No` aparece con **151 registros**, equivalentes al **61.89%** del total. La categoría `Yes` aparece con **93 registros**, equivalentes al **38.11%**.

Esto significa que hay más registros de mesas no fumadoras que de mesas fumadoras.

Como en los casos anteriores, esta tabla describe la composición del dataset. No indica todavía si un grupo consume más, deja más propina o tiene cuentas más altas. Para responder esas preguntas necesitaremos combinar categorías con variables numéricas.

## Repetir el análisis en otra columna

Para completar esta primera revisión de frecuencias, vamos a aplicar el mismo procedimiento a la columna `sex`.

La pregunta es la misma que antes:

**¿Cuántos registros hay en cada categoría y qué porcentaje representa cada una?**

In [18]:
frecuencia_sexo = pd.DataFrame({
    "frecuencia": df["sex"].value_counts(),
    "porcentaje": df["sex"].value_counts(normalize=True) * 100
})

frecuencia_sexo["porcentaje"] = frecuencia_sexo["porcentaje"].round(2)

frecuencia_sexo

,frecuencia,porcentaje
sex,,
Male,157,64.34
Female,87,35.66


La tabla muestra la distribución de la columna `sex`.

En el dataset, la categoría `Male` aparece con **157 registros**, equivalentes al **64.34%** del total. La categoría `Female` aparece con **87 registros**, equivalentes al **35.66%**.

Esto indica que, dentro de este dataset, hay más registros asociados a `Male` que a `Female`.

Nuevamente, esta salida describe cantidades y proporciones. No permite concluir, por sí sola, que un grupo consuma más o deje más propina. Para eso habría que combinar esta variable categórica con columnas numéricas como `total_bill` o `tip`.

## Cierre del capítulo

En este capítulo trabajamos con columnas categóricas.

A diferencia de las columnas numéricas, que pueden resumirse con promedios, medianas, mínimos o máximos, las columnas categóricas suelen analizarse contando cuántas veces aparece cada valor.

Vimos tres formas de leer estas columnas:

- **frecuencias absolutas**, que indican la cantidad de registros por categoría;
- **frecuencias relativas**, que indican qué proporción del total representa cada categoría;
- **porcentajes**, que expresan esas proporciones de una manera más fácil de interpretar.

También construimos tablas de frecuencia que combinan cantidad de registros y porcentaje. Esta forma de presentación permite describir mejor la composición del dataset.

Sin embargo, contar categorías no alcanza para responder preguntas más analíticas, como qué día tiene mayor facturación promedio, qué horario deja mejores propinas o si un grupo tiene cuentas más altas que otro.

Para responder ese tipo de preguntas necesitaremos combinar categorías con variables numéricas. Ese será el próximo paso en el análisis.